### ISSUE TITLE
Worse compression ratio compare to ZLIB and Intel's IPP version of ZLIB

### ISSUE THREAD

**Issue by showvick** [intel/isa-l#197]
We are running performance tests comparing ISAL to ZLIB and IPP variant of ZLIB compression level 1. In general we see significant reduction in CPU and about the same compression ratio (0-5%). 

But with the attached dataset we see about 15% difference in size. Wanted to understand why that is so? Is that due to the randomness in the data? Or ISAL doesn't try further back in the window? I believe ISAL also uses 32 KB window size. 

Below is how we ISAL 
 ```
        isal_deflate_stateless_init(&isalstream);
        isalstream.end_of_stream = 1;
        isalstream.flush = NO_FLUSH;
        isalstream.next_in = (Bytef *)cmpinstart_p;
        isalstream.avail_in = bytestocompress;
        isalstream.next_out = (Bytef *)cmpoutstart_p;
        isalstream.avail_out = length;
        isalstream.gzip_flag = IGZIP_ZLIB;

        isalstream.level = 2;
        isalstream.level_buf = <buffer of below size>
        isalstream.level_buf_size = ISAL_DEF_LVL2_DEFAULT;

        zret = isal_deflate_stateless(&isalstream);
```

We use compression level 1 with ZLIB and Intel's IPP variant of ZLIB. 

I have attached the dataset
db2.bin --> Uncompressed data file
db2_ipp.bin --> Compression file with Intel's IPP version
db2_isal.bin --> Compressed file with ISAL
db2_zlib.bin --> Compressed file with ZLIB

Thanks in advance. Your help is greatly appreciated. 

[data.zip](https://github.com/intel/isa-l/files/7865384/data.zip)

**Comment by gbtucker** [comment]
ISA-L deflate compression does use 32KB window size by default.  Compression is of course a trade-off between speed and compression ratio.  Our levels don't always match ratios exactly with other implementations on every dataset but we seem to be way above the curve even on this dataset.  For example for db2.bin, I'm getting on ISA-L level 3 a higher compression ratio (201813 bytes out) vs zlib level 4 (202633 bytes out) and still over 5x faster.

**Comment by showvick** [comment]
Thanks Greg. Appreciate it. 
The difference between level 2 and level 3 is substantial. Would be nice if there was something in between. 

Can something be done to improve the compression ratio at level 2 for this dataset? Would using a bigger buffer  ISAL_DEF_LVL2_EXTRA_LARGE instead of ISAL_DEF_LVL2_DEFAULT help?

**Comment by gbtucker** [comment]
The throughput difference between level 2 and level 3 is much lower on modern Xeon processors.  (Xeon are better at level 3 comparatively).  Changing the level buffer size might not help much in this case.  zlib also has a big jump in compression ratio between levels 3 and 4 likely because, like our 2 and 3, changing strategies finds more longer matches.

**Comment by showvick** [comment]
Thanks Greg. Can you please elaborate a little on modern Xeon processors? Is it starting from a  CPU type like skylake/cascade lake? Or CPU that supports AVX512/AVX2? Can we check it programmatically? 

We tried level 3 compression on the same test, there was a significant difference in throughput. The compression ratio was much much better but significant difference in throughput. The CPU improvement dropped from 70% to 13% versus Intel IPP version of ZLIB. So, about a 50% drop comparing level 2 and level 3. 

Below is the CPU we used for the test
```
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                96
On-line CPU(s) list:   0-95
Thread(s) per core:    2
Core(s) per socket:    24
Socket(s):             2
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 85
Model name:            Intel(R) Xeon(R) Gold 5220R CPU @ 2.20GHz
Stepping:              7
CPU MHz:               2201.000
CPU max MHz:           2201.0000
CPU min MHz:           1000.0000
BogoMIPS:              4389.65
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              1024K
L3 cache:              36608K
NUMA node0 CPU(s):     0-95
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm flush_l1d md_clear ibrs_enhanced constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch ida arat epb invpcid_single pln pts dtherm intel_pt rsb_ctxsw spec_ctrl arch_capabilities ssbd ibpb stibp kaiser tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 cqm_llc cqm_occup_llc pku ospke
```

**Comment by gbtucker** [comment]
Yes, for level 3 I see a big improvement starting at skylake due to the availability of AVX512 and the throughput gap seems to be subsequently narrowing between level 2 and 3.

https://www.intel.com/content/www/us/en/developer/articles/technical/intel-isa-l-and-intel-integrated-performance-primitives-zlib-solutions.html has some more comparisons between ISA-L with IPP deflate compression.

**Comment by showvick** [comment]
Thanks for sharing.
